### ISSUE TITLE
zetas- Ram crashes dask (GPU/CPU)

### ISSUE THREAD

**Issue by kh-abd-kh** [dask/dask#11442]
Hi, Thank you for your  hard work.

I am evaluating 2d  zetas(twisted with harmonic polynomials): the lemniscate.
just sum over x,y for (x**4 - 6. * x**2 * y**2 + y**4)/(x**2+y**2)**4
I can do this in numba/cython/c/cuda up to 10**12 terms (1 000 000,1 000 000)

Now, with dask,  it crahes my WSL completly even for small values (100 000,100 000)

Here the GPU code (the CPU is just Client)

from dask_cuda import LocalCUDACluster
from dask.distributed import Client

############################################

`import numpy as np
import cupy as cp

import dask.array as da
import dask

from dask_cuda import LocalCUDACluster
from dask.distributed import Client

cluster = LocalCUDACluster()
client = Client(cluster)
client

def zet(x,y):
    return ((x+1)**4 -6.*((x+1)**2)*((y+1)**2) + (y+1)**4)/((x+1)**2+(y+1)**2)**4

with dask.config.set({"array.backend": "cupy"}):
    xg = da.fromfunction(zet,shape=(50000,50000),dtype="float64")

%time xg.sum().compute()
`

############################################

It crahes for higher values of shape (100000,100000). Also when I put some higher values
of chunks (i have 6G gpu), it crashes.

The RAM consumption increases then it craches and the WSL terminal disapper

system:

└─$ pip list|grep dask
dask                         2024.9.0
dask-cuda                    24.12.0a7
dask-cudf-cu12               24.12.0a188
dask-expr                    1.1.14
rapids-dask-dependency       24.12.0a6
#################################
cuda-python                  12.6.0
cudf-cu12                    24.12.0a188
cupy-cuda12x                 13.3.0
dask-cuda                    24.12.0a7
dask-cudf-cu12               24.12.0a188

libcudf-cu12                 24.12.0a188
numba-cuda                   0.0.17
nvidia-cublas-cu12           12.6.3.3
nvidia-cuda-cupti-cu12       12.6.80
nvidia-cuda-nvcc-cu12        12.6.77
nvidia-cuda-runtime-cu12     12.6.77
nvidia-cudnn-cu12            9.5.0.50
nvidia-cufft-cu12            11.3.0.4
nvidia-cusolver-cu12         11.7.1.2
nvidia-cusparse-cu12         12.5.4.2
nvidia-nccl-cu12             2.23.4
nvidia-nvcomp-cu12           4.0.1.0
nvidia-nvjitlink-cu12        12.6.77
pycuda                       2024.1.2
pylibcudf-cu12               24.12.0a188
pynvjitlink-cu12             0.3.0
rmm-cu12                     24.12.0a16

Over kali/Ubuntu22  WSL 
Device 0: "NVIDIA GeForce GTX 1660 Ti with Max-Q Design"
  CUDA Driver Version / Runtime Version          12.7 / 12.6
  CUDA Capability Major/Minor version number:    7.5
  Total amount of global memory:                 6144 MBytes (6442123264 bytes)

My system: 
Ryzen 7/4800  16GB RAM
Win 11

Any help will very much appreciated.

Thank you.

**Comment by kh-abd-kh** [comment]
sorry: the editor refactor my code and removed spaces.

**Comment by phofl** [comment]
@dask/gpu

**Comment by pentschev** [comment]
If I understand correctly from your code, it seems the data you're generating is `shape=(50000,50000),dtype="float64"`, is that correct? This is 20GB where you have a 6GB GPU, that will not fit.

**Comment by kh-abd-kh** [comment]
But chunks will divide and distribute
first: i put 
with dask.config.set({"array.backend": "cupy"}):
    xg = da.fromfunction(zet,chunks=(5000,5000),shape=(50000,50000),dtype="float64")

But then i found , if i remove chunks , dask wwill chose smaller chunks.


Both cases, with chunks=(5000,5000) and without chunks (leave dask to chose suitable usually smaller),
both cases crachs once  i go for higher shape.

I understand chunks is the important not shape.

**Comment by pentschev** [comment]
Yes, it will divide and distribute if you have multiple workers where to distribute that, in your case you have one worker (one GPU) so it cannot be distributed anywhere else. The same applies for your CPU, if you don't have at least 20GB free it will crash.

If you have small enough chunks, you _may_ be able to still get it working if you have at least enough disk space to fit the problem if you specify `--memory-limit` (for CPU) and `--device-memory-limit` (for GPU), but it will of course be much slower as data has to be moved around.

**Comment by pentschev** [comment]
> > dask/gpu
> 
> sorry, what does it mean?!

Just tagging the GPU group to respond, since you were asking about a GPU use case.

**Comment by kh-abd-kh** [comment]
There are two Rams GPU-RAM, i haven't seen problems,
 CPU-RAM makes me for me whenit shouldn't.
i have used up to chunks=(15000,15000), it offloads the GPU completley(=6GB)


The data is  cupy GPU array, Dask should not generate CPU array.
JUST generate cupy (chunk) GPU array, sum them over the GPU and collect 
the whole sum latter.

I understand that Dask can handle Terra bytes of data by suitable chunksize them
over the GPU/CPU.

**********************************************************
OK, If i am doing something terribley wrong.
How to sum this series 
sum_{x,y=1}^{x,y=1 000 000} {(x^4 -6*x^2 y^2 + y^4)/(x^2+y^2)^4}
over the gpu, by chunksize it up say chunks=(10000,10000).

Thank you, you should add explicit example in the maual for  da.fromfunction.
IT  is top important for math/physics.

**Comment by rjzamora** [comment]
>I understand that Dask can handle Terra bytes of data by suitable chunksize them
over the GPU/CPU.

Thanks for the issue @kh-abd-kh  - Yes, you should be able to process this amount of data using a single/GPU as long as a single "chunk" fits comfortably in memory (probably ~1/8 single-GPU memory or less).

It seems like the `fromfunction` algorithm/implementation is attempting to hold a lot of data in memory at once. I will try to learn a bit more about this function (how it works) before I say whether or not there is a good solution (or work-around).

**Comment by kh-abd-kh** [comment]
Thank you guys for reformating up my code.

another implicit issues that i faced,
a)
    xg = da.fromfunction(zet,shape=(50000,50000),dtype="float64")
generated numpy array. shouldn't be  cupy array with  LocalCUDACluster()
by default ?! instead of forcing it by
with dask.config.set({"array.backend": "cupy"}):

b)
Also, if i used big chunks bigger than my gpu-ram, it simply and correctly
quit with a logical message(out of memory, trying to allocate ...  )
but here no message not abort and complete cpu-ram crashes

c)
And please put some gpu-cupy examples other than the random generated ones.
and emphasis the "dask.config.set({"array.backend": "cupy"}):"
case.

Thank you a lot. sorry, for being more math/phys  oriented.
But, i find dask should be useful and more usable with math.

**Comment by rjzamora** [comment]
>shouldn't be cupy array with LocalCUDACluster() by default ?! instead of forcing it by with dask.config.set({"array.backend": "cupy"}):

No, the data API (i.e. Dask Array) and the API you use for cluster deployment (Dask-CUDA) are unrelated. Dask-CUDA just does some extra work for you to make sure your workers each have access to a distinct GPU (Dask-CUDA workers also have other nice features for host->device spilling).

With that said, I do think it would make sense for there to be a coordination between Worker-resource types and the data-API backend. I'm personally very interested in working on this, but I'd expect it to take time.

>Also, if i used big chunks bigger than my gpu-ram, it simply and correctly
quit with a logical message(out of memory, trying to allocate ... )
but here no message not abort and complete cpu-ram crashes

The out-of-memory error is not coming from Dask when cupy is used. Dask/Distributed will simply bubble-up whatever errors the backend produces. Maybe cupy (or maybe RMM) is throwing a more useful error than Numpy is in this case?

>And please put some gpu-cupy examples other than the random generated ones.
and emphasis the "dask.config.set({"array.backend": "cupy"}):"
case.

Thanks for the suggestion - You are absolutely correct that the GPU-related documentation is lacking. The best way to improve the documentation is to contribute new examples (or raise issues like you're doing right now - thanks!).

**Comment by kh-abd-kh** [comment]
Thank you, very much.
To be honest, I found it is difficult to figure out what is dask  and what is dask-cuda 
and their hidden inter-relation with numpy-cupy.
It took a couple of hours googling and trying out different codes to find  this
with dask.config.set({"array.backend": "cupy"}):

you, guys, should include it in the first line of your dask -gpu 

docs.dask.org/en/stable/gpu.html

and, also

examples.dask.org/index.html

and more proper in 

docs.rapids.ai/api/dask-cuda/nightly/index.htm

Many-thanks.

**Comment by rjzamora** [comment]
@kh-abd-kh - I had a chance to experiment with your reproducer today. I don't have access to a 6GB GPU, but I tried to artificially reduce the memory limit to "mimic" your system.

I think the key to good performance in your use case is [`jit_unspill=`True`](https://docs.rapids.ai/api/dask-cuda/stable/spilling/#jit-unspill). I seem to get reasonable performance for `shape=(100_000, 100_000)` with the following `LocalCUDACluster` code:

```python
cluster = LocalCUDACluster(
    device_memory_limit=0.6,  # I set to "4GB" to "mimic" your system
    rmm_pool_size=0.8,  # I set to "5GB" to "mimic" your system
    jit_unspill=True,  # This seems to be the most important argument!
)
```

Performance improves a bit with larger chunks sizes. For example, using `chunks=(8192, 8192)` will reduce the number of tasks and improve GPU utilization compared to `chunks=(4096, 4096)`. However, we do want the chunks to be small enough for the Dask-CUDA worker to easily spill and unspill the data if needed.

Let me know if this helps at all. Thanks again for engaging!

**Comment by kh-abd-kh** [comment]
You are THE MAN.
jit_unspill=True
solved  the problem. No more cpu-ram problems ot crashes.

I shall return to you later with a full report.
Now, I am playing with the 100_000, 200_000. Of course somoone like you should have noticed
that this is the complex generalization of the euler zeta. Now, I am playing also, with the 4d quaternionic
case. I shall return to you with a breif report including some nice graphs.
And, I promise to write up some few page so take whatever you like to your manuals.

Amazing, we have dask-gpu-solved what Gauss had calculated ~1793 and remained unknown
untill Klein published Gauss private diary in 1913!

I shall be back and sorry for making you work in the weekend.

PS: what is the equivalent magic trick for dask-cpu?

**Comment by charlesbluca** [comment]
> To be honest, I found it is difficult to figure out what is dask and what is dask-cuda and their hidden inter-relation with numpy-cupy. It took a couple of hours googling and trying out different codes to find this with dask.config.set({"array.backend": "cupy"}):

I'm sorry to hear this 😕 digging through the documentation you called out, I can definitely see how the setting of `array.backend` (or `dataframe.backend`) would be unclear.

> docs.dask.org/en/stable/gpu.html

It looks like the code snippets used here don't reflect more recent work around backend dispatching; I'd argue that they could probably be updated to resemble the snippets in the [collection backend docs](https://docs.dask.org/en/stable/how-to/selecting-the-collection-backend.html), so maybe something like:

```python
import dask
import dask.array as da

x = da.ones(10, chunks=(5,))  # Creates numpy-backed collection

with dask.config.set({"array.backend": "cupy"}):
    y = da.ones(10, chunks=(5,))  # Get cupy-backed collection
    gpu_x = x.to_backend()  # Moves numpy data to cupy
```

And something similar for dataframes.

> examples.dask.org/index.html

This one is a little tougher, since my understanding is that the examples here are also intended to be run on a Binder spun off from [dask/dask-examples](https://github.com/dask/dask-examples), which imposes some [limitations](https://github.com/dask/dask-examples?tab=readme-ov-file#contributing) that I could imagine make it difficult to trivially contribute compelling modern GPU examples to the repo. That being said, it could be worth revisiting how often people spin up the Binder versus checking the HTML versions of the notebooks at examples.dask.org to get a sense of if these limitations are still worthwhile.

> docs.rapids.ai/api/dask-cuda/nightly/index.html

Looking at this page, I feel like the problematic statement is perhaps

> While Distributed can be used to leverage GPU workloads through libraries such as [cuDF](https://docs.rapids.ai/api/cudf/stable/), [CuPy](https://cupy.dev/), and [Numba](https://numba.pydata.org/), Dask-CUDA offers several unique features unavailable to Distributed...

In that it doesn't necessarily make clear _how_ cuDF and CuPy can be used directly with Distributed - maybe a better way to phrase this that emphasizes the collection backend dispatching would be

> While Distributed can be used to leverage GPU workloads by setting the [collection backend](https://docs.dask.org/en/stable/how-to/selecting-the-collection-backend.html) to a GPU library that interfaces with Dask, such as [cuDF](https://docs.rapids.ai/api/cudf/stable/) or [CuPy](https://cupy.dev/)...

That being said, I could see someone following the link to our collection backend docs with intent to learn more about cuDF getting confused since there isn't an example of setting `dataframe.backend` there, so might be good to add a similar example to the one above for dataframes:

```python
import dask
import dask.dataframe as dd

ddf1 = ...  # Creates pandas-backed collection

with dask.config.set({"dataframe.backend": "cudf"}):
    ddf2 = ...  # Get cudf-backed collection
    gpu_ddf1 = ddf1.to_backend()  # Moves pandas data to cudf
```

(Don't mean to hijack the issue 😉 just trying to form some substantive action items from your valid comments)

**Comment by kh-abd-kh** [comment]
I am sorry to disapoint you and i am grateful for your reply. Thank you for helping me learning dask.

My backgroung = physics/math- multi-genus (Riemann Surface) solitons solutions
I have solved this problem already using cython/numba.openACC(nvc++)
and i hoped dask will be better or more uniform.
i can't figure out the correct relation between chunk size and performance.
It is more or less trial/error from my part.
In the complex case, the chunks=(8192,8192),shape=(250_000,250_000)
leads to 961 chunks in 6 layers. it finished in 15 min.
the 4d case is more interesting
chunks=(96,96,96,96),shape=(384,384,384,384) leads to 256 chunks in 10 layers.
I haven't yet tried the 8d which is more exciting Numba fails to give correct results!!!!

Numba and Cython, in general,  can do better, the gpu case nvc++ is much better than numba/cython.
It is clear i am missing many things. about how to fine-tune dask and making faster, making it works
is not enough for me.
Thank you for your constructive comment, i shall read it carefully.

PS: i am still stuck with cpu case.

**Comment by kh-abd-kh** [comment]
My main motiviation is to have a dask hyprid systam
a) wsl ububtu 22, where i am running dask inside dask shell,i.e i plan
to use dask with sage ( https://www.sagemath.org ) and pari (  https://pari.math.u-bordeaux.fr/download.html )
b) wsl kali, where cuda tools and nvdia hpc sdk are installed, and i am running dask-cuda

i.e and i hope somehow latter to make them talk and offloading the cpu and the gpu at the same time.


For now i am grateful to rick for solving the gpu part of the problem. The dask-cpu still open.

It seems now, there are some dask performance issue. Why numba/cython/nvc++ are much faster?
Is there some way to make dask faster?

**Comment by kh-abd-kh** [comment]
A little word about the manuals, third-world countries are not online i.e it is
always nice to have a full pdf version.

**Comment by kh-abd-kh** [comment]
It seems for the dask-cpu,i am running against unmanaged memory problems.
https://blog.dask.org/2021/11/02/choosing-dask-chunk-sizes
https://www.coiled.io/blog/tackling-unmanaged-memory-with-dask
Thank you guys for your help.

**Comment by rjzamora** [comment]
>PS: what is the equivalent magic trick for dask-cpu?

I don't think there is a cpu equivalent for `jit_unspill=True`. However, I don't think I fully understand why that option is helping so much on Dask-CUDA workers anyway (@madsbk might have some insight).

For the cpu case (with `LocalCluster`), it seems a bit easier to tune the memory management ([docs here](https://distributed.dask.org/en/latest/worker-memory.html#worker-memory-management)) with `threads_per_worker=1`. This way, you can set a specific memory limit on each worker, and that worker will only execute one task at a time (this is essentially what `threads_per_worker=1` means). For example, if you have 16GB of RAM, and want to use 4 workers, you might try something like:

```python
dask.config.set({"array.backend": "numpy"})
cluster = LocalCluster(
    n_workers=4, # Set an explicit worker count (just to be sure of "total" memory limit)
    memory_limit="4GB",  # This is the limit per-worker. You could also set this a bit lower to make sure you stay under 16GB
    threads_per_worker=1,  # This reduces concurrency, but makes memory-management easier to tune
)
```

**Comment by kh-abd-kh** [comment]
Thank you again, i shall try tonight and come back later.

**Comment by kh-abd-kh** [comment]
Wow, you gave me the hint, in the page refered up in your reply. It is mentioned that the cpu problem
is over Linux. But i have win11 and dask-cpu is installed. 
Now, I ran 7 worker from win11 with ram limit 2G + dask-cuda from wsl kali
Amaizing total cpu-ram is under 10G (when i start it is 6G) and both of them at the same time  done the 4d
case 
 xc = da.fromfunction(zet4,chunks=(64,64,64,64),shape=(256,256,256,256),dtype="float64")   
for chunk 128.0 MB and total array 32GB. 

Now, i shall their performance and higher values for the 2d and 4d and see how the 8d can
work 8 sums amounts to N^8 for N=100 shall be 10 Thousands Trillion term each float64. Only could nvc++
could finished this one in 30-40 mins.

i shall return later.

**Comment by kh-abd-kh** [comment]
i forget to mention the numbers
for the gpu finished in 
CPU times: user 18.9 s, sys: 4.1 s, total: 23 s
Wall time: 2min 7s
array(-0.00785531)

for the cpu finished in 
CPU times: total: 38 s
Wall time: 2min 57s
-0.007855309429094112

the gpu alone 
CPU times: user 3.99 s, sys: 1.09 s, total: 5.07 s
Wall time: 1min 30s
array(-0.00785531)

**Comment by kh-abd-kh** [comment]
for my configuration, i use :
jupyter lab --no-browser
one from win11 and another from wsl/kali
and then i collect them from win11 edge two running kernels with different two different dask
one cpu and the other is gpu. Now, the cpu is just 6 workers to give enough room for the gpu.

**Comment by kh-abd-kh** [comment]
Now, is it possible to use dask to distribute the same calculation between the two client A+B?
A- dask-cpu over win11
B- dask-gpu over wsl/kali 
Since the dask-gpu seems to be faster can i do it (40/60)%  load?

**Comment by madsbk** [comment]
> Now, is it possible to use dask to distribute the same calculation between the two client A+B?

No, not out-the-box :/ Dask-CUDA assigns one worker per GPU. 
We have discussed a mixed execution model, like you suggest, but I don't think it was implemented.

**Comment by kh-abd-kh** [comment]
Yes, after few trials, i understand, it needs higher level of abstractation.
Dask translates the code into gpu and cpu.  it is  needed some fusion of numpy/cupy
which looks like possible.

But, you know in my case it is easy, just one line of code
with dask.config.set({"array.backend": "cupy"}):
or 
with dask.config.set({"array.backend": "numpy"}):

They are two different clients: A) cpu B) gpu

Actualy, with dask miracle: one win11 and another wsl/kali/ubuntu 
we ,at least, doubled the performance.

can i distribute the code (40cpu/60gpu)%  between A) and B) ?

**Comment by kh-abd-kh** [comment]
By the way i forget to mention that when i ran two clients
one from win11 and anther from wsl/kali . It is ok no ports conflict.
In the the earlier case ws/unbuntu22 and wsl/kali, there were some ports conflicts but
no harms.
